### PR TITLE
NSM manager registry chain + memory reg fixes.

### PR DESCRIPTION
**Pull request by haiodo**
+ Chain element for NSMgr registry chain
+ Fixes for bulk register tail and stuff.

Signed-off-by: Andrey Sobolev <haiodo@gmail.com>

### CODE REVIEW

File: pkg/registry/nsmgr/server.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+
+	"github.com/golang/protobuf/ptypes/empty"
+	"github.com/networkservicemesh/api/pkg/api/registry"
+
+	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
+	"github.com/networkservicemesh/sdk/pkg/tools/serialize"
+)
+
+type nsmRegistryServer struct {
+	registry.NetworkServiceRegistryServer
```
Comment (denis-tingajkin) [inline pkg/registry/nsmgr/server.go:31]: Why do we need this field?

File: pkg/registry/nsmgr/server.go [hunk 2]
```diff
@@ -0,0 +1,100 @@
...
+	serialize.Executor
+	manager *registry.NetworkServiceManager
+}
+
+// NewServer - construct a new NSM aware NSE registry server
+// Server is using latest NSM registration and update endpoints accordingly.
+func NewServer(manager *registry.NetworkServiceManager) registry.NetworkServiceRegistryServer {
+	return &nsmRegistryServer{
+		manager: manager,
```
Comment (denis-tingajkin) [inline pkg/registry/nsmgr/server.go:40]: Do we allow the nil manager here?

File: pkg/registry/memory/nse_memory_server.go [hunk 3]
```diff
@@ -32,6 +32,15 @@ type memoryNetworkServeRegistry struct {
...
 func (m *memoryNetworkServeRegistry) RegisterNSE(ctx context.Context, registration *registry.NSERegistration) (*registry.NSERegistration, error) {
+	nseRegistration, err := m.processRegistration(registration)
+	if err != nil {
+		return nseRegistration, err
+	}
+
+	return next.NetworkServiceRegistryServer(ctx).RegisterNSE(ctx, registration)
+}
+
+func (m *memoryNetworkServeRegistry) processRegistration(registration *registry.NSERegistration) (*registry.NSERegistration, error) {
```
Comment (denis-tingajkin) [inline pkg/registry/memory/nse_memory_server.go:43]: Do we really need this method? Can we in bulk server call `m.RegisterNSE(...)` ?

File: pkg/registry/nsmgr/server.go [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+	}
+}
+
+// RegisterNSE - register NSE with current manager associated.
+func (n *nsmRegistryServer) RegisterNSE(ctx context.Context, registration *registry.NSERegistration) (*registry.NSERegistration, error) {
+	// update manager to be proper one.
+	<-n.AsyncExec(func() {
+		registration.NetworkServiceManager = n.manager
+	})
```
Comment (denis-tingajkin) [inline pkg/registry/nsmgr/server.go:49]: Why do we need to use synchronization?

File: pkg/registry/nsmgr/server.go [hunk 5]
```diff
@@ -0,0 +1,100 @@
...
+	result, err := next.NetworkServiceRegistryServer(ctx).RegisterNSE(ctx, registration)
+	if result != nil {
```
Comment (denis-tingajkin) [inline pkg/registry/nsmgr/server.go:51]: Should we check `err` here?

File: pkg/registry/nsmgr/server.go (hunk 1 above)
Comment (haiodo) [reply pkg/registry/nsmgr/server.go:31]: removed.

File: pkg/registry/nsmgr/server.go (hunk 2 above)
Comment (haiodo) [reply pkg/registry/nsmgr/server.go:40]: NewServer is not return error, so I could not return error here. But I think manager should be passed.

File: pkg/registry/nsmgr/server.go (hunk 4 above)
Comment (haiodo) [reply pkg/registry/nsmgr/server.go:49]: Yes, we don't. Removed.

File: pkg/registry/memory/nse_memory_server.go (hunk 3 above)
Comment (haiodo) [reply pkg/registry/memory/nse_memory_server.go:43]: We can't since it will loose stream capability.

File: pkg/registry/nsmgr/server.go (hunk 5 above)
Comment (haiodo) [reply pkg/registry/nsmgr/server.go:51]: thanks.

File: pkg/registry/common/localbypass/server.go [hunk 6]
```diff
@@ -21,42 +21,56 @@ package localbypass
...
 	"github.com/networkservicemesh/api/pkg/api/registry"
-
-	"github.com/networkservicemesh/sdk/pkg/registry/core/next"
 )
 
-type localBypassRegistry struct {
-	sockets localbypass.SocketMap
+// LocalBypass - return interafce to return localbypass endpoints URIs
+type LocalBypass interface {
+	LoadEndpoint(endpointName string) (*url.URL, bool)
```
Comment (denis-tingajkin) [inline pkg/registry/common/localbypass/server.go:33]: [minor] Should we rename it to `LoadPass` ?

File: pkg/registry/core/next/discovery_server.go [hunk 7]
```diff
@@ -40,7 +40,9 @@ type nextDiscoveryServer struct {
 func NewWrappedDiscoveryServer(wrapper DiscoveryServerWrapper, servers ...registry.NetworkServiceDiscoveryServer) registry.NetworkServiceDiscoveryServer {
 	rv := &nextDiscoveryServer{servers: make([]registry.NetworkServiceDiscoveryServer, 0, len(servers))}
 	for _, c := range servers {
-		rv.servers = append(rv.servers, wrapper(c))
+		if c != nil {
```
Comment (denis-tingajkin) [inline pkg/registry/core/next/discovery_server.go:43]: Do we ignore with this check that SDK developer doing something wrong?

File: pkg/networkservice/common/localbypass/server_test.go [hunk 8]
```diff
@@ -23,6 +23,8 @@ import (
 	"net/url"
 	"testing"
 
+	localbypass_reg "github.com/networkservicemesh/sdk/pkg/registry/common/localbypass"
```
Comment (denis-tingajkin) [inline pkg/networkservice/common/localbypass/server_test.go:26]: I would suggest this name
```suggestion
	registrylocalbypass "github.com/networkservicemesh/sdk/pkg/registry/common/localbypass"
```

File: pkg/registry/core/next/discovery_server.go (hunk 7 above)
Comment (haiodo) [reply pkg/registry/core/next/discovery_server.go:43]: There was no such check, but it simplify usage of this kind of wrapper servers a lot.

File: pkg/networkservice/common/localbypass/server_test.go (hunk 8 above)
Comment (haiodo) [reply pkg/networkservice/common/localbypass/server_test.go:26]: Thanks.

File: pkg/registry/common/localbypass/server.go (hunk 6 above)
Comment (haiodo) [reply pkg/registry/common/localbypass/server.go]: Thanks.

**Comment by edwarnicke** [comment]
Can this be broken into smaller pieces?  There seem to be several things in motion here.

**Pull request closed without merging**